patches/node/fix_isurl_implementation.patch

```diff
@@ -0,0 +1,33 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Charles Kerr <charles@charleskerr.com>
+Date: Thu, 20 Jul 2023 14:18:19 -0500
+Subject: fix: isURL() implementation
+
+Modify Node's lib/internal/url.js isURL() to return the correct value for
+URLs created both inside and outside of Node.
+
+The isURL() impl in `main` is Electron-safe but can't be backported in
+isolation because it relies on other changes from 18 to main. But we can
+safely get there by trying the 18 version first (catching Node URLs),
+then the `main` version (catching Electron URLs).
+
+More background w/upstream links at
+https://github.com/electron/electron/pull/39154#issuecomment-1644433388
+
+This patch can be removed if upstreamed.
+
+diff --git a/lib/internal/url.js b/lib/internal/url.js
+index 36604c9de2179b5f6941dd9a7af66cdb03facd7f..27c3d1c1909ead49c4384d6d53d60c40115e9542 100644
+--- a/lib/internal/url.js
++++ b/lib/internal/url.js
+@@ -592,7 +592,9 @@ ObjectDefineProperties(URLSearchParams.prototype, {
+  * @returns {self is URL}
+  */
+ function isURL(self) {
+-  return self != null && ObjectPrototypeHasOwnProperty(self, context);
++  if (self != null && ObjectPrototypeHasOwnProperty(self, context))
++    return true; // a Node URL
++  return Boolean(self?.href && self.origin); // another impl's URL
```

Comment:
The origin getter is extremely slow, and might call c++ depending on the URL. I recommend using another method and validating it against !== undefined for better performance.

Reply:
@anonrig even though the origin getter is slow, it's the one that's present in Node 18.17, which is the version that Electron's currently tracking. I totally get what you're saying, but deviating from the version we're targeting makes me a little nervous.

Do you have a suggestion on what would be a safe + more efficient alternative to the version in the patch above?

Reply:
On Node.js 18 we are using the following (ref: https://github.com/nodejs/node/blob/v18.x-staging/lib/internal/url.js#L595)

```
function isURL(self) {
  return self != null && ObjectPrototypeHasOwnProperty(self, context);
}
```

On Node 20, we are using: (Ref: https://github.com/nodejs/node/blob/main/lib/internal/url.js#L762C3-L762C74)

```
return Boolean(self?.href && self.protocol && self.auth === undefined);
```

After Node 18, we moved to using private properties for better performance and avoid calling `isURL`. But there is still an edge case a contributor found (ref: https://github.com/nodejs/node/pull/48928/files#diff-26e6b1e2a90040b1ce20e7f305b2ff3140b483f863ee2573b4c6af9b696ad31eR762), that does not properly distinguish the legacy url and the `URL` class. This PR is currently in queue to get merged.

```
return Boolean(self?.href && self.protocol && self.auth === undefined && self.path === undefined);
```

My suggestion is to go with the soon-to-be-merged solution.

Reply:
@anonrig right, I understand that about 18 and 20. I hadn't seen that edge case PR so :bowing_man: good catch about needing to pick up that new change too.

The issue with using either 20 as-is or the soon-to-be-merged solution as-is is the getters that it calls, e.g.

```js
function isURL(self) {
  return Boolean(self?.href && self.protocol && self.auth === undefined && self.path === undefined);
}
```

The problem is backporting that impl to 18 causes an infinite loop because of the property getters that it calls, e.g. this href getter in 18:

```js
  get href() {
    if (!isURL(this))
      throw new ERR_INVALID_THIS('URL');
    return this[context].href;
  }
```

So isURL() unconditionally calls the href getter, and the href getter unconditionally calls isURL(), so there is an infinite loop. This is what I meant [upthread](https://github.com/electron/electron/pull/39154#issuecomment-1644513147) about the infinite loop but I should've been more explicit about the problem.

Reply:
Ah you're right @ckerr. In that particular scenario, I recommend using Object.getOwnPropertyDescriptor which does not trigger the getters?

Such as:

```
Welcome to Node.js v20.5.0.
Type ".help" for more information.
> const a = { get hello() { return 'world' } }
undefined
> Object.getOwnPropertyDescriptor(a, 'hello').enumerable
true
```